### PR TITLE
Convert some is_ot_match calls to simple comparisons

### PR DESCRIPTION
**Pull request by jbytheway**
#### Summary
None

#### Purpose of change
The `is_ot_match` function is a genericised way of checking certain properties of an `oter_id`.  It makes sense when the type of check is loaded from JSON, but when the type of check is either `exact` or `type` and the value compared against is known, using this function is redundant; the same can be achieved via a simple comparison.

This has a few benefits:
* It's less code.
* It's more performant (in particular, no `std::string` allocations, which were showing up as quite expensive when profiling overmap generation).
* It detects and reports an error message when we compare against an invalid id, helping avoid code rot.

#### Describe the solution
Write a clang-tidy check to detect such cases and automatically convert them to simple comparisons.

#### Describe alternatives you've considered
Making the changes by hand only in the cases where profiling indicated the necessity.

#### Testing
`clang-tidy` and unit tests.

#### Additional context
I plan to check the performance improvement when I get the chance.  Will report back here when I do.

In order for this check to work reasonably well, I had to add/improve other checks first.  That's what inspired #52411, #52691, #52726, and #52766.

### CODE REVIEW

**Comment by jbytheway** [comment]
Results of my performance testing:
Here are before and after flame graphs of the performance of `generate_sub` (one of the main functions used in overmap generation), both sampled in the same (unpublished) unit test:
ot_match-profile-before
After:
ot_match-profile-after
You can perhaps see that the fraction of the `generate_sub` runtime used by `is_ot_match` and by string construction has decreased (although not by as much as I hoped).
`is_ot_match` used to be 21%, now it's 13%.  string construction has gone from 21% to 12%.

Overall, the fraction of time overmap generation is spending in `generate_sub` decreased from 17.3% to 16.1% (although that change might well be within the margin or error).

So, that seems good.  Unfortunately, the overall time for the unit test appears to have actually got worse (from 104s to 109s); but I'm hoping that's due to unrelated changes, since I also rebased on latest master between these tests.

In conclusion, I think this change had the desired effect, and I stand by it as a good change, but its overall impact was a little underwhelming.  I continue to look for other opportunities to improve mapgen performance.  I think we should be able to eliminate all string construction from this function.
